### ISSUE TITLE
Transformation/Peek operators in `StreamMessage`

### ISSUE THREAD

**Issue by trustin** [line/armeria#3097]
Related StackOverflow question: https://stackoverflow.com/questions/64188193/using-response-interceptor-to-access-returned-headers

We could add some additional operators to `StreamMessage` so that a user can easily transform or peek the stream:

```java
HttpResponse res1 = res0.peek(obj -> { ... });
HttpResponse res2 = res1.map(obj -> { ... });
HttpResponse res3 = res2.flatMap(obj -> { ... });

// Certain types only:
res.peek(ResponseHeaders.class, headers -> { ... })
   .map(HttpData.class, data -> { ...});
```

We might want to benchmark other reactive frameworks such as Reactor and Mutiny or even implement their API.

**Comment by injae-kim** [comment]
@trustin , @ikhoon nim, can I try this issue? 😄

**Comment by ikhoon** [comment]
Yes. Please go ahead.

**Comment by ikhoon** [comment]
JFYI, `map` operation was implemented at #3624 
```
HttpResponse res2 = res1.map(obj -> { ... });
```

**Comment by injae-kim** [comment]
Thanks! I'll take this issue.

First PR creation target date is `2021-11-24 (Next Wed)` !
